### DISCUSSION TITLE
Home Assistant auto-discovery

### DISCUSSION THREAD

**Discussion by flacjacket** [teslamate-org/teslamate#1599] [Ideas]
One feature that Home Assistant exposes is the ability to use [discovery](https://www.home-assistant.io/docs/mqtt/discovery/) to find sensors that are published via MQTT. Using MQTT discovery means that users would not need to define all of the sensors manually in the `sensors.yaml` file, and furthermore all of the sensors would be associated with the same Device in Home Assistant. All that is needed for autodiscovery is to publish messages like `<prefix>/sensor/<nodeid>/<objectid>/config` that contains the data that would otherwise be present in the yaml config files. These could be sent as retained messages at start-up. One possible way to configure this would be to set a configuration variable `HOME_ASSISTANT_PREFIX` that would configure the autodiscovery prefix to use.

I am not too familiar with Elixir or where to look to add this sort of functionality, but would be happy to help figure out the right format for the messages.

**Comment by sillyfrog** [comment]
I wanted something similar (including control of the vehicle), so I have written this: https://github.com/sillyfrog/teslabuddy

It is another container, but it reads the info from TeslaMate (via MQTT), and publishes it for Home Assistant, including aggregating the GPS info into a single JSON payload so it'll work with auto-discovery.

**Comment by mjdyson** [comment]
Has there been any thought on this? Personally, I think this would be a great addition to TeslaMate.

Might it be better to implement in addition to existing MQTT approaches? Im sure there are many users using MQTT a certain way already.

Feels like there needs to be an attribute map to identify the prefixes required for each of the sensors. It also allows for sensor types (temperature, location, etc).

An example of how this presents to the user is as follows: 

![image](https://user-images.githubusercontent.com/4598068/178106244-a54cdb9c-d503-4d52-a79a-b90f7c3aeb1d.png)
![image](https://user-images.githubusercontent.com/4598068/178106253-f4e99ee6-a641-4c04-9264-61dfaf93939d.png)

**Comment by nebhale** [comment]
I'd love to know whether @adriankumpf is interested in an external contribution for this.

In an attempt to iron out all of the payload contents (there's a fair bit of `value_template`-ing necessary to get it to all work) I [wrote a script](https://github.com/nebhale/teslamate-discovery) that creates all the payloads and publishes them.  Now that I have them, I'd be up for creating a PR that contributes the functionality in Elixir, but it's a big enough job that I don't want to start unless he's open to the contribution.

**Reply by cwanja** [reply]
I think Adrian has been pretty vocal that if there is enough community desire for it, he is willing to include it. So I would say he’d be open to accepting a PR. We have collaborating on a few things I doubt he had in mind. So he is certainly willing to accept external contributions.

**Reply by mjdyson** [reply]
Superb! I started trying to piece together some value templates for the various sensors, but I'm glad you've found a way to do this @nebhale. 

I've been trying to get my head around how it all works, and how it might be implemented. I'm not an experienced dev, but i figured...
1) create a mapping file for each of the sensors to the associated value template so the value can be looked up via a key/value pair.
2) create a setting boolean for enabling publishing this way
3) create an initial MQTT message with /config suffix (with retain?)
4) adjust the function that generates the MQTT message to include the value template lookup, potentially within the line
`topic =
      ["teslamate", namespace, "cars", car_id, key]`

For part 4, I believe it's here.
https://github.com/adriankumpf/teslamate/blob/master/lib/teslamate/mqtt/pubsub/vehicle_subscriber.ex
Lines 79-86 appear to be where the actual MQTT message is constructed.

**Reply by sillyfrog** [reply]
I would also suggest looking at my code as well here: https://github.com/sillyfrog/teslabuddy
It includes how to do the HA config message (having retain on that is important for when you restart HA), I couldn't get GPS to work with the output from TM, so needed to aggregate it all into a single message (the docs for GPS especially are very poor - at least when I did it).

Good luck!

**Reply by nebhale** [reply]
Thanks for all the responses!  To clarify, I've got a firm grasp on the Home Assistant side of things (including the [synthesized `device_tracker`](https://github.com/nebhale/teslamate-discovery/blob/main/teslamate-discovery.sh#L73-L76)).  You can run [the script](https://github.com/nebhale/teslamate-discovery) against your Home Assistant install and a new Device and [66 Entities](https://github.com/nebhale/teslamate-discovery/blob/main/teslamate-discovery.sh#L261-L336) will magically appear.

My question is more whether I should dive into the codebase and create a PR that sends those 66 MQTT messages for each car.  It's a pretty sizable change which means a pretty big investment from @adriankumpf to guide it to completion, even if I write the code.  As a general courtesy I'm looking for a 👍 before obligating him to that work 😇.

**Comment by Gerben321** [comment]
Any update on this? A native solution would be great.

**Comment by brianseaders** [comment]
i know its not mqtt related but the truth is I have not figured out how to send a DM on GitHub I was wondering what the level of difficulty would likely be in updating the amcrest integration so that it can function in GUI instead of being a legacy integration and if anybody's actually working on that already cuz I'm kind of thinking about learning how to do it myself I'm really really tired of ymal config and I have about 30 cameras that I use with home assistant. I a very aware of the fact that it would be a monumental undertaking for someone at my skill level and much harder than continuing to use the legacy integration but i want it...

**Comment by nebhale** [comment]
Now implemented: https://github.com/teslamate-org/teslamate/releases#release-v4.1.0
